### PR TITLE
[3.8] Namespace document

### PR DESCRIPTION
**Pull request by fastslack**
Namespace for document classes

### Summary of Changes

Move /libraries/joomla/document to /libraries/src/Joomla/CMS/Document

### Testing Instructions

Apply patch and try to create one article

### Expected result

No errors

### CODE REVIEW

**Comment by mbabker** [comment]
This has test issues.

```sh
Error displaying the error page: Application Instantiation Error: include_once(/home/travis/build/joomla/joomla-cms/libraries/joomla/document/html/renderer/component.php): failed to open stream: No such file or directory
```

**Comment by fastslack** [comment]
@mbabker I removed that files because are duplicated and are not needed any more. Look at the files, there is only mapping to deprecated files.

https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/html/renderer/component.php

We can solve compatibilities issues in classmap.php file. 

https://github.com/fastslack/joomla-cms/blob/namespace-document/libraries/classmap.php#L191

Maybe we should remove that tests

**Comment by mbabker** [comment]
We don't need the format in the short class section for the renderers.  So for example instead of `Joomla\CMS\Document\Renderer\Html\RendererHtmlComponent` we can do something like `Joomla\CMS\Document\Renderer\Html\ComponentRenderer`.

**Comment by mbabker** [comment]
OK, one more pedantic change, then there's one concern we need to quickly look over.  For the document classes, can you change it to `<format>Document` instead of `Document<format>`?

The concern - we have some files here that have multiple classes in them.  I think this would be a good time to move them out to separate files.

File: libraries/src/Joomla/CMS/Document/Document.php [hunk 1]
```diff
@@ -285,24 +286,29 @@ public static function getInstance($type = 'html', $attributes = array())
...
 				{
-					JLog::add('Non-autoloadable JDocument subclasses are deprecated, support will be removed in 4.0.', JLog::WARNING, 'deprecated');
+					\JLog::add('Non-autoloadable Document subclasses are deprecated, support will be removed in 4.0.', \JLog::WARNING, 'deprecated');
 				}
 				// Default to the raw format
 				else
 				{
 					$ntype = $type;
-					$class = 'JDocumentRaw';
+					$class = 'DocumentRaw';
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/Document.php]: Namespace is missing, or?

File: libraries/src/Joomla/CMS/Document/Document.php [hunk 2]
```diff
@@ -1165,20 +1171,25 @@ public function _getTab()
...
+		if (!class_exists($class))
+		{
+			$class = 'JDocumentRenderer' . ucfirst($this->getType()) . ucfirst($type);
+		}
 
 		if (!class_exists($class))
 		{
 			// "Legacy" class name structure
-			$class = 'JDocumentRenderer' . $type;
+			$class = 'DocumentRenderer' . $type;
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/Document.php]: Namespace is missing, or?

File: libraries/src/Joomla/CMS/Document/OpensearchDocument.php [hunk 3]
```diff
@@ -226,11 +230,11 @@ public function addImage(JOpenSearchImage $image)
 }
 
 /**
- * JOpenSearchUrl is an internal class that stores the search URLs for the OpenSearch description
+ * OpenSearchUrl is an internal class that stores the search URLs for the OpenSearch description
  *
  * @since  11.1
  */
-class JOpenSearchUrl
+class OpenSearchUrl
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/OpensearchDocument.php]: This classe should be moved to its own file.

File: libraries/src/Joomla/CMS/Document/OpensearchDocument.php [hunk 4]
```diff
@@ -264,11 +268,11 @@ class JOpenSearchUrl
 }
 
 /**
- * JOpenSearchImage is an internal class that stores Images for the OpenSearch Description
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/OpensearchDocument.php]: This classe should be moved to its own file.

File: libraries/src/Joomla/CMS/Document/OpensearchDocument.php [hunk 5]
```diff
@@ -264,11 +268,11 @@ class JOpenSearchUrl
...
+ * OpenSearchImage is an internal class that stores Images for the OpenSearch Description
  *
  * @since  11.1
  */
-class JOpenSearchImage
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/OpensearchDocument.php]: This classe should be moved to its own file.

File: libraries/src/Joomla/CMS/Document/XmlDocument.php [hunk 6]
```diff
@@ -78,7 +81,7 @@ public function getName()
 	 *
 	 * @param   string  $name  Document name
 	 *
-	 * @return  JDocumentXml instance of $this to allow chaining
+	 * @return  JXmlDocument instance of $this to allow chaining
```
Comment (laoneo) [inline libraries/src/Joomla/CMS/Document/XmlDocument.php]: J needs to be removed.

**Review by laoneo — CHANGES_REQUESTED** [review]

**Comment by laoneo** [comment]
Thanks!

File: libraries/src/Joomla/CMS/Document/Document.php [hunk 7]
```diff
@@ -1165,15 +1171,20 @@ public function _getTab()
...
 	 *
 	 * @since   11.1
 	 * @throws  RuntimeException
 	 */
 	public function loadRenderer($type)
 	{
-		// New class name format adds the format type to the class name
-		$class = 'JDocumentRenderer' . ucfirst($this->getType()) . ucfirst($type);
+		// Determine the path and class
+		$class = __NAMESPACE__ . '\\Renderer\\' . ucfirst($this->getType()) . '\\Renderer' . ucfirst($this->getType()) . ucfirst($type);
```
Comment (mbabker) [inline libraries/src/Joomla/CMS/Document/Document.php:1182]: If I'm reading this line right it's not building the correct namespaced class names.

File: libraries/src/Joomla/CMS/Document/FeedDocument.php [hunk 8]
```diff
@@ -242,11 +246,11 @@ public function addItem(JFeedItem $item)
 }
 
 /**
- * JFeedItem is an internal class that stores feed item information
+ * FeedItem is an internal class that stores feed item information
  *
  * @since  11.1
  */
-class JFeedItem
+class FeedItem
```
Comment (mbabker) [inline libraries/src/Joomla/CMS/Document/FeedDocument.php:253]: This and the other classes in this file should move to separate files now.

**Comment by mbabker** [comment]
Fixed the items I pointed out and merged to staging.

**Pull request closed without merging**